File: manual/code/test_suites/cpp/FunctionBuildTest/tests/test_bitarray_extend.cpp
#include <gtest/gtest.h>
#include "../src/function_bitarray_extend.cpp"

class TestBitarrayExtender : public ::testing::Test {
protected:
    BitarrayExtender extender;
};

TEST_F(TestBitarrayExtender, TestExtendWithValidBits) {
    std::string result = extender.extend_bits("101", "010");
    EXPECT_EQ(result, "101010");
}

TEST_F(TestBitarrayExtender, TestInvalidInputNonBinary) {
    EXPECT_THROW(extender.extend_bits("101", "102"), std::invalid_argument);
}

TEST_F(TestBitarrayExtender, TestNonStringInput) {
    EXPECT_THROW(extender.extend_bits("101", "abc"), std::invalid_argument);
}

TEST_F(TestBitarrayExtender, TestExtendWithEmptyString) {
    std::string result = extender.extend_bits("101", "");
    EXPECT_EQ(result, "101");
}

TEST_F(TestBitarrayExtender, TestExtendEmptyStringWithBits) {
    std::string result = extender.extend_bits("", "010");
    EXPECT_EQ(result, "010");
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: manual/code/test_suites/python/python/test_bitarray_extend.py

import unittest
from function_bitarray_extend import BitarrayExtender

class TestExtendBitsFunction(unittest.TestCase):

    def setUp(self):
        self.extender = BitarrayExtender()

    def test_extend_with_valid_bits(self):
        result = self.extender.extend_bits("101", "010")
        self.assertEqual(result, "101010")

    def test_invalid_input_non_binary(self):
        with self.assertRaises(ValueError):
            self.extender.extend_bits("101", "102")

    def test_non_string_input(self):
        with self.assertRaises(TypeError):
            self.extender.extend_bits(101, "010")

    def test_extend_with_empty_string(self):
        result = self.extender.extend_bits("101", "")
        self.assertEqual(result, "101") 

    def test_extend_empty_string_with_bits(self):
        result = self.extender.extend_bits("", "010")
        self.assertEqual(result, "010")  

if __name__ == "__main__":
    unittest.main()
